fast64_internal/f3d/flipbook.py

```diff
@@ -0,0 +1,329 @@
+import bpy, re
+from typing import List, Any, Callable
+from bpy.utils import register_class, unregister_class
+from bpy.app.handlers import persistent
+from .f3d_material import all_combiner_uses, update_tex_values_manual, iter_tex_nodes, TextureProperty
+from ..utility import prop_split
+
+
+class TextureFlipbook:
+    def __init__(self, name: str, exportMode: str, textureNames: List[str]):
+        self.name = name
+        self.exportMode = exportMode
+        self.textureNames = textureNames
+
+
+def flipbook_data_to_c(flipbook: TextureFlipbook):
+    newArrayData = ""
+    for textureName in flipbook.textureNames:
+        newArrayData += textureName + ",\n"
+    return newArrayData
+
+
+def flipbook_to_c(flipbook: TextureFlipbook, isStatic: bool):
+    newArrayData = "void* " if not isStatic else "static void* "
+    newArrayData += f"{flipbook.name}[] = {{ "
+    newArrayData += flipbook_data_to_c(flipbook)
+    newArrayData += " };"
+    return newArrayData
+
+
+def flipbook_2d_to_c(flipbook: TextureFlipbook, isStatic: bool, count: int):
+    newArrayData = "void* " if not isStatic else "static void* "
+    newArrayData += f"{flipbook.name}[][{len(flipbook.textureNames)}] = {{ "
+    for i in range(count):
+        newArrayData += "{ " + flipbook_data_to_c(flipbook) + " },\n"
+    newArrayData += " };"
+    return newArrayData
+
+
+def usesFlipbook(
+    material: bpy.types.Material,
+    flipbookProperty: Any,
+    index: int,
+    checkEnable: bool,
+    checkFlipbookReference: Callable[[str], bool],
+) -> bool:
+    texProp = getattr(material.f3d_mat, "tex" + str(index))
+    if all_combiner_uses(material.f3d_mat)["Texture " + str(index)] and texProp.use_tex_reference:
+        return (
+            checkFlipbookReference is not None
+            and checkFlipbookReference(texProp.tex_reference)
+            and (not checkEnable or flipbookProperty.enable)
+        )
+    else:
+        return False
+
+
+class ImagePointerProperty(bpy.types.PropertyGroup):
+    image: bpy.props.PointerProperty(type=bpy.types.Image)
+    name: bpy.props.StringProperty(name="Name", default="gImage")
+
+
+def drawTextureArray(
+    layout: bpy.types.UILayout, textureArray: bpy.types.CollectionProperty, index: int, exportMode: str
```

Comment:
That type should still be technically correct, I think? If not descriptive? I feel like it would be better to leave it as opposed to not having a type at all.

Reply:
yeah I agree it's a shame there's no good choice to type it, one option would be to add our own type, maybe in utility.py, like `CollectionProperty = Any  # collection prop as defined by using bpy.props.CollectionProperty` maybe